### PR TITLE
Named enclaves

### PR DESCRIPTION
**Pull request by raulmldv**
### User friendly names for enclaves

Introducing a new option for the `run-enclaves` command allowing users to set custom names for running enclaves instead of using only the ID. The name is set following the option `--enclave-name` and will be uniquely assigned to each of the running enclaves. In case of name conflicts, an index will be concatenated to the initial name, incrementing until the name becomes unique. This action is optional and if no name is provided, the EIF name will take its place.
The name will be part of the output of the following commands: `run-enclave`, `describe-enclaves` and `terminate-enclave`.

Commands that use the ID for targeting a specific enclave, now have the option to address the enclave by its name. So the commands `console` and `terminate-enclave` will have the option to use `--enclave-name` instead of `--enclave-id`.

Unique name assigning is done by the CLI process by broadcasting a request for all the running enclaves' names. The new name is checked agains all the others and saved in the EnclaveManager or changed if needed (with the index added) . Name addressing for the other commands is done by sending, again, to all te running enclaves, the name provided by the user. The matching enclave will send its ID and the command will be executed as before, using the ID.

### Command outputs

```
$ ./nitro-cli run-enclave --cpu-count 2 --memory 256 --enclave-cid 16 --eif-path hello.eif --enclave-name testName --debug-mode
Start allocating memory...
Started enclave with enclave-cid: 16, memory: 256 MiB, cpu-ids: [1, 9]
{
  "EnclaveName": "testName",
  "EnclaveID": "i-0584873aac8fea23b-enc17b9c558602bf9f",
  "ProcessID": 652683,
  "EnclaveCID": 16,
  "NumberOfCPUs": 2,
  "CPUIDs": [
    1,
    9
  ],
  "MemoryMiB": 256
}

$ ./nitro-cli describe-enclaves
[
  {
    "EnclaveName": "testName",
    "EnclaveID": "i-0584873aac8fea23b-enc17b9c558602bf9f",
    "ProcessID": 652683,
    "EnclaveCID": 16,
    "NumberOfCPUs": 2,
    "CPUIDs": [
      1,
      9
    ],
    "MemoryMiB": 256,
    "State": "RUNNING",
    "Flags": "DEBUG_MODE",
    "Measurements": {
      "HashAlgorithm": "Sha384 { ... }",
      "PCR0": "3eaa317a778ee207fab894f6b42c5545ddef8bd905da8cc2f22cc59cc83fe0c980a411e5f5445f110f26d5c0d6f39ee4",
      "PCR1": "aca6e62ffbf5f7deccac452d7f8cee1b94048faf62afc16c8ab68c9fed8c38010c73a669f9a36e596032f0b973d21895",
      "PCR2": "ea436e643c31f2891161803e3f22084a12d38f807ef733379cadf789ec85d51e3bb787f48f3cad0c291fbc82fe7ca27d"
    }
  }
]

$ ./nitro-cli console --enclave-name testName
Connecting to the console for enclave 16...
Successfully connected to the console.
[    0.000000] Linux version 4.14.177-104.253.amzn2.x86_64 (mockbuild@ip-10-0-1-32) (gcc version 7.3.1 20180712 (Red Hat 7.3.1-6) (GCC)) #1 SMP Fri May 1 02:01:13 UTC 2020
..........

```
In the scenario that two enclaves could run simultaneously, a second run command with the exact name will result:

```
$ ./nitro-cli run-enclave --cpu-count 2 --memory 256 --enclave-cid 16 --eif-path hello.eif --enclave-name testName --debug-mode
Start allocating memory...
Started enclave with enclave-cid: 16, memory: 256 MiB, cpu-ids: [1, 9]
{
  "EnclaveName": "testName_1",
  "EnclaveID": "i-0584873aac8fea23b-enc17b9c55other_id",
  "ProcessID": 652683,
  "EnclaveCID": 16,
  "NumberOfCPUs": 2,
  "CPUIDs": [
    1,
    9
  ],
  "MemoryMiB": 256
}
```

```
$ ./nitro-cli terminate-enclave --enclave-name testName
Successfully terminated enclave i-0584873aac8fea23b-enc17b9c558602bf9f.
{
  "EnclaveName": "testName",
  "EnclaveID": "i-0584873aac8fea23b-enc17b9c558602bf9f",
  "Terminated": true
}
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/common/commands_parser.rs [hunk 1]
```diff
@@ -392,6 +396,16 @@ fn debug_mode(args: &ArgMatches) -> Option<bool> {
...
 }
 
+/// Parse the enclave name from the command-line arguments.
+fn parse_enclave_name(args: &ArgMatches) -> NitroCliResult<Option<String>> {
+    let enclave_name = if let Some(enclave_name) = args.value_of("enclave-name") {
+        Some(enclave_name.to_string())
+    } else {
+        None
+    };
+    Ok(enclave_name)
```
Comment (costisin) [inline src/common/commands_parser.rs:427]:
```suggestion
    Ok(args.value_of("eif-path").map(|e| e.to_string()))
```

File: src/common/json_output.rs [hunk 2]
```diff
@@ -76,8 +80,13 @@ pub struct DescribeOutput {
 
 impl DescribeOutput {
     /// Creates new describe output from available
-    pub fn new(describe_info: EnclaveDescribeInfo, build_info: Option<EnclaveBuildInfo>) -> Self {
+    pub fn new(
+        enclave_name: Option<String>,
```
Comment (costisin) [inline src/common/json_output.rs:84]: Why don't you include enclave_name in the EnclaveDescribeInfo struct?

File: src/enclave_proc/commands.rs [hunk 3]
```diff
@@ -48,12 +48,24 @@ pub fn run_enclaves(
         .map_err(|e| e.add_subaction("Failed to construct CPU information".to_string()))?
         .get_cpu_config(args)
         .map_err(|e| e.add_subaction("Failed to get CPU configuration".to_string()))?;
+
+    let enclave_name = match args.enclave_name.clone() {
```
Comment (costisin) [inline src/enclave_proc/commands.rs:52]: Is cloning necessary?

File: src/lib.rs [hunk 4]
```diff
@@ -424,6 +424,41 @@ pub fn get_all_enclave_names() -> NitroCliResult<Vec<String>> {
...
+            NitroCliErrorEnum::EnclaveNamingError
+        ));
+    } else if objects.len() > 1 {
+        return Err(new_nitro_cli_failure!(
+            "Conflicting enclave names have been found.".to_string(),
+            NitroCliErrorEnum::EnclaveNamingError
+        ));
+    }
+
+    Ok(objects[0].clone())
```
Comment (costisin) [inline src/lib.rs:459]: Why not just moving?

File: src/lib.rs [hunk 5]
```diff
@@ -591,14 +637,23 @@ macro_rules! create_app {
                             .long("enclave-id")
                             .takes_value(true)
                             .help("Enclave ID, used to uniquely identify an enclave")
-                            .required(true),
+                            .required(true)
```
Comment (costisin) [inline src/lib.rs:640]: You are looking for `required_unless("enclave-name")` which requires an argument unless other argument `enclave-name` is found

File: src/lib.rs [hunk 6]
```diff
@@ -424,6 +424,41 @@ pub fn get_all_enclave_names() -> NitroCliResult<Vec<String>> {
...
+        return Err(new_nitro_cli_failure!(
+            "No enclave matched the given name.".to_string(),
+            NitroCliErrorEnum::EnclaveNamingError
+        ));
+    } else if objects.len() > 1 {
+        return Err(new_nitro_cli_failure!(
+            "Conflicting enclave names have been found.".to_string(),
+            NitroCliErrorEnum::EnclaveNamingError
+        ));
+    }
```
Comment (costisin) [inline src/lib.rs:457]: Maybe you could rewrite like this so there is no duplicating code
```suggestion
    if (objects.len() != 1) {
        return Err(new_nitro_cli_failure!(
            match objects.len() {
                0 => "No enclave matched the given name.".to_string(),
                2.. => "Conflicting enclave names have been found.".to_string(),
            },
            NitroCliErrorEnum::EnclaveNamingError
        ));
    }
```

File: src/common/commands_parser.rs (hunk 1 above)
Comment (raulmldv) [reply src/common/commands_parser.rs:427]: will change with "enclave-name" instead of "eif-path"

File: src/common/json_output.rs (hunk 2 above)
Comment (raulmldv) [reply src/common/json_output.rs:84]: I avoided changing the returned type by the enclave for keeping the backwards compatibility. I will look for a way of refactoring to avoid confusion.

File: src/enclave_proc/commands.rs (hunk 3 above)
Comment (raulmldv) [reply src/enclave_proc/commands.rs:52]: could use to_string later and borrowing here

File: src/lib.rs (hunk 4 above)
Comment (raulmldv) [reply src/lib.rs:459]: Cannot move out of vector index, could use remove() instead.

File: src/common/commands_parser.rs (hunk 1 above)
Comment (costisin) [reply src/common/commands_parser.rs:427]: yeah, sorry, i meant "enclave-name"

File: src/lib.rs [hunk 7]
```diff
@@ -155,6 +157,40 @@ pub fn build_from_docker(
...
+            .set_action("Get Enclaves Name".to_string())
+    })?;
+
+    let mut idx = 0;
+    let mut result_name = enclave_name.clone();
+    names.sort();
+
+    // If duplicates are found, add index to name eg. testName -> testName_1 -> testName_2 ..
+    for name in names {
+        if result_name == name {
```
[the remark this answers is not in the document]
Comment (raulmldv) [reply src/lib.rs:185]: Once an enclave is assigned a name with the index, it will be added to this array. I realised that it is not guaranteed that I will find all the existing names with indexes in consecutive order in the sorted array (eg. list is: name, nameother, name_1 ...) so I should check this in another way.

File: src/common/json_output.rs [hunk 8]
```diff
@@ -35,20 +39,27 @@ pub struct EnclaveDescribeInfo {
     #[serde(rename = "Flags")]
     /// The bit-mask which provides the enclave's launch flags.
     pub flags: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(flatten)]
+    /// Build measurements containing PCRs
+    pub build_info: Option<EnclaveBuildInfo>,
```
Comment (petreeftime) [inline src/common/json_output.rs:45]: This seems to be part of another patch.

File: src/common/json_output.rs [hunk 9]
```diff
@@ -35,20 +39,27 @@ pub struct EnclaveDescribeInfo {
...
     pub fn new(
+        enclave_name: Option<String>,
         enclave_id: String,
         enclave_cid: u64,
         cpu_count: u64,
         cpu_ids: Vec<u32>,
         memory_mib: u64,
         state: String,
         flags: String,
+        build_info: Option<EnclaveBuildInfo>,
```
Comment (petreeftime) [inline src/common/json_output.rs:59]: This seems to be part of another patch.

File: src/enclave_proc/commands.rs [hunk 10]
```diff
@@ -48,12 +48,24 @@ pub fn run_enclaves(
...
+
+    let enclave_name = match &args.enclave_name {
+        Some(enclave_name) => enclave_name,
+        None => {
+            return Err(new_nitro_cli_failure!(
+                "Failed to set enclave name in the manager".to_string(),
+                NitroCliErrorEnum::FileOperationFailure
+            ))
+        }
+    };
```
Comment (petreeftime) [inline src/enclave_proc/commands.rs:60]: Perhaps I'm missing something, but does that mean that enclave_name is not an optional argument? I don't think it should be a required argument. Perhaps setting a useful default is better (eg, the enclave ID or https://github.com/fnichol/names).

**Review by petreeftime — COMMENTED** [review]
How backwards compatible is this change? If I start an enclave on the current published version and then upgrade the tooling to this, will it still work, or will something crash?

File: src/common/json_output.rs (hunk 8 above)
Comment (raulmldv) [reply src/common/json_output.rs:45]: > How backwards compatible is this change? If I start an enclave on the current published version and then upgrade the tooling to this, will it still work, or will something crash?

I tested backwards compatibility and developed the feature in this way. The output of the current published version will be missing the additional field of the describe output which will be interpreted by the new CLI process as being `None`, so nothing crashes.

File: src/enclave_proc/commands.rs (hunk 10 above)
Comment (raulmldv) [reply src/enclave_proc/commands.rs:60]: There is a default value to which the name is set in case the argument is missing, the name of the EIF (test.eif -> test). Still, this field is kept as an `Option` because at the moment of parsing the arguments it could be missing.

File: src/common/json_output.rs (hunk 9 above)
Comment (raulmldv) [reply src/common/json_output.rs:59]: This is added so the output of the `describe-enclaves` includes the new information. Should this be part of another commit or another PR?

File: src/lib.rs [hunk 11]
```diff
@@ -155,6 +157,32 @@ pub fn build_from_docker(
...
+/// Creates new enclave name
+///
+/// Requests the names of all running instances and checks the
+/// occurrence of the chosen name for the new enclave.
+pub fn new_enclave_name(run_args: RunEnclavesArgs, names: Vec<String>) -> NitroCliResult<String> {
+    let enclave_name = match run_args.enclave_name {
+        Some(enclave_name) => enclave_name,
+        None => {
+            // Get name of EIF file from path eg. path/to/eif/hello.eif -> hello
+            let path_split: Vec<&str> = run_args.eif_path.split(&['/', '.'][..]).collect();
```
Comment (axlprv) [inline src/lib.rs:169]: We should not assume that all the EIF paths have the `.eif` extension as we do not enforce this rule anywhere.

File: src/common/json_output.rs (hunk 9 above)
Comment (petreeftime) [reply src/common/json_output.rs:59]: Whichever you think makes more sense. I prefer smaller/more PRs.

File: src/lib.rs (hunk 11 above)
Comment (raulmldv) [reply src/lib.rs]: Solved to treat the case of no extension present.

File: src/common/json_output.rs (hunk 9 above)
Comment (raulmldv) [reply src/common/json_output.rs:59]: I'll keep that in mind for future PRs, for now I added a new commit because it makes more sense with the suggestion of @costisin in this PR.

**Review by axlprv — APPROVED** [review]

File: src/lib.rs [hunk 12]
```diff
@@ -419,6 +419,39 @@ pub fn get_all_enclave_names() -> NitroCliResult<Vec<String>> {
     Ok(objects)
 }
 
+/// Sends the name to all the running enclaves and expects a response
+/// with the ID of the one that uniquely matched
```
Comment (petreeftime) [inline src/lib.rs:423]: What guarantees that names are unique? I haven't seen any check in run-enclave that ensures this.

Comment (petreeftime) [reply src/lib.rs:423]: I've searched a bit more and finally found it.

**Review by petreeftime — APPROVED** [review]
